### PR TITLE
Improve documentation for `kubernetes.io/enforce-mountable-secrets`

### LINKED ISSUE #40727: Improve documentation for `kubernetes.io/enforce-mountable-secrets` annotation on `ServiceAccount`
The documentation for the `kubernetes.io/enforce-mountable-secrets` annotation on `ServiceAccount` consists of [this paragraph on the **Well-Known Labels, Annotations and Taints** page](https://kubernetes.io/docs/reference/labels-annotations-taints/#enforce-mountable-secrets):

> The value for this annotation must be true to take effect. This annotation indicates that pods running as this service account may only reference Secret API objects specified in the service account's secrets field.

This leaves unclear the details of the actual effect of the annotation. By checking [the code](https://github.com/kubernetes/kubernetes/blob/release-1.26/plugin/pkg/admission/serviceaccount/admission.go#L302) I was able to determine that setting this annotation to `true` causes the following enforcements:

(1) `Secret`s mounted as `volumes` must appear in the `ServiceAccount`'s `secrets`
(2) `Secret`s referenced in `envFrom` in `containers` or `initContainers` must appear in the `ServiceAccount`'s `secrets`
(3) `Secret`s referenced in `imagePullSecrets` of `Pod`s must appear in the `ServiceAccount`'s `imagePullSecrets`

(1) is implied by the name of the annotation, but the lack of any mention of (2) or (3) in the documentation forced me to check the code to ensure that it wasn't only enforcing *mounting* of `Secret`s as the name implies. The documentation also mentions only the `secrets` field of the `ServiceAccount`, yet the enforcement also takes into account the `ServiceAccount`'s `imagePullSecrets` field (which the `Pod`'s `imagePullSecrets` are checked against).

The documentation would be improved by describing the actual enforcement behaviour that this feature implements.

This feature may also benefit from being mentioned in the [Good practices for Kubernetes Secrets](https://kubernetes.io/docs/concepts/security/secrets-good-practices/) document.

### PR DESCRIPTION
**Pull request by chansuke**
Fixes #40727 

Update the description on `ServiceAccount`.

1.Expanded the description to clearly explain the actual enforcement behavior, highlighting not just the mounting of secrets, but also their references in envFrom and imagePullSecrets.
2. Emphasized the need to ensure all referenced secrets are correctly specified in the ServiceAccount to avoid potential issues

### CODE REVIEW

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 1]
```diff
@@ -493,8 +493,15 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
...
 The value for this annotation must be **true** to take effect.
-This annotation indicates that Pods running as this ServiceAccount may only reference
-Secret API objects specified in the ServiceAccount's `secrets` field.
+When set to **true**, this annotation enforces the following rules for Pods running as this ServiceAccount:
+
+1. Secrets mounted as volumes must be listed in the ServiceAccount's `secrets` field.
+2. Secrets referenced in `envFrom` for containers or `initContainers` must also be listed in the ServiceAccount's `secrets` field.
+3. Secrets referenced in a Pod's `imagePullSecrets` must be present in the ServiceAccount's `imagePullSecrets` field.
+
+While the annotation's name suggests it only concerns the mounting of secrets,
```
Comment (tengqm) [inline content/en/docs/reference/labels-annotations-taints/_index.md:502]:
```suggestion
While the annotation's name suggests it only concerns the mounting of Secrets,
```

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 2]
```diff
@@ -493,8 +493,15 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
...
+its enforcement also extends to other ways Secrets are used in the context of a Pod.
+Therefore, it is crucial to ensure that all the referenced secrets are correctly specified in the ServiceAccount.
```
Comment (tengqm) [inline content/en/docs/reference/labels-annotations-taints/_index.md:504]:
```suggestion
Therefore, it is crucial to ensure that all the referenced Secrets are correctly specified in the ServiceAccount.
```

File: content/en/docs/reference/labels-annotations-taints/_index.md (hunk 1 above)
Comment (chansuke) [reply content/en/docs/reference/labels-annotations-taints/_index.md]: Thank you for your review.

**Comment by sftim** [comment]
We avoid Markdown in commit messages

/retitle Improve documentation for kubernetes.io/enforce-mountable-secrets

**Review by sftim — COMMENTED** [review]
I would like https://kubernetes.io/docs/concepts/configuration/secret/ to mention this annotation. Perhaps also https://kubernetes.io/docs/concepts/security/secrets-good-practices/?

@chansuke are you willing to edit the Secret concept to also mention this annotation?

**Comment by sftim** [comment]
We could also edit https://kubernetes.io/docs/concepts/security/service-accounts/ to mention this annotation and link to the section about its effect.

**Comment by chansuke** [comment]
@sftim Yes, I'd like to edit the documentation too.Can I edit this on this PR?

File: content/en/docs/concepts/security/service-accounts.md [hunk 3]
```diff
@@ -196,6 +196,31 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
 You can also use TokenRequest to obtain short-lived tokens for your external application.
 {{< /note >}}
 
+### Enforcing Mountable Secrets using Annotations {#enforce-mountable-secrets}
+
+Kubernetes provides an annotation called `kubernetes.io/enforce-mountable-secrets` that you can add to your ServiceAccounts.
+When this annotation is applied, the service account's secrets can only be mounted on specified types of resources, enhancing the security posture of your cluster.
+
+#### How to Add the Annotation
```
Comment (tengqm) [inline content/en/docs/concepts/security/service-accounts.md:204]: This doesn't need to be a subsection.

File: content/en/docs/concepts/security/service-accounts.md [hunk 4]
```diff
@@ -196,6 +196,31 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+kind: ServiceAccount
+metadata:
+  annotations:
+    kubernetes.io/enforce-mountable-secrets: "true"
+  name: my-serviceaccount
+  namespace: my-namespace
+```
+
+#### Effect on the Cluster
+
```
Comment (tengqm) [inline content/en/docs/concepts/security/service-accounts.md:219]: Remove this.

File: content/en/docs/concepts/security/service-accounts.md [hunk 5]
```diff
@@ -196,6 +196,31 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+When this annotation is set to true, the Kubernetes control plane ensures that the secrets belonging to this ServiceAccount are only mountable on the resource types specified in the annotation value.
```
Comment (tengqm) [inline content/en/docs/concepts/security/service-accounts.md:220]:
```suggestion
When this annotation is set to "true", the Kubernetes control plane ensures that
the Secrets from this ServiceAccount are only mountable on the resource kinds
specified in the annotation value.
```

File: content/en/docs/concepts/security/service-accounts.md [hunk 6]
```diff
@@ -196,6 +196,31 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+For more information about the authentication process, refer to
+[kubernetes.io/enforce-mountable-secrets {#enforce-mountable-secrets}](/docs/reference/labels-annotations-taints/_index.md#kubernetesioenforce-mountable-secrets-enforce-mountable-secrets).
```
Comment (tengqm) [inline content/en/docs/concepts/security/service-accounts.md:222]:
```suggestion
[the documentation on this annotation](/docs/reference/labels-annotations-taints/#enforce-mountable-secrets).
```

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 7]
```diff
@@ -530,8 +530,15 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
 Used on: ServiceAccount
 
 The value for this annotation must be **true** to take effect.
-This annotation indicates that Pods running as this ServiceAccount may only reference
-Secret API objects specified in the ServiceAccount's `secrets` field.
+When set to **true**, this annotation enforces the following rules for Pods running as this ServiceAccount:
```
Comment (tengqm) [inline content/en/docs/reference/labels-annotations-taints/_index.md:533]:
```suggestion
When set to "true", this annotation enforces the following rules for Pods running
as this ServiceAccount:
```

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 8]
```diff
@@ -530,8 +530,15 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
...
+
+1. Secrets mounted as volumes must be listed in the ServiceAccount's `secrets` field.
+2. Secrets referenced in `envFrom` for containers or `initContainers` must also be listed in the ServiceAccount's `secrets` field.
+3. Secrets referenced in a Pod's `imagePullSecrets` must be present in the ServiceAccount's `imagePullSecrets` field.
```
Comment (tengqm) [inline content/en/docs/reference/labels-annotations-taints/_index.md:537]:
```suggestion
1. Secrets mounted as volumes must be listed in the ServiceAccount's `secrets` field.
1. Secrets referenced in `envFrom` for containers or `initContainers` must also
   be listed in the ServiceAccount's `secrets` field.
1. Secrets referenced in a Pod's `imagePullSecrets` must be present in the
   ServiceAccount's `imagePullSecrets` field.
```

File: content/en/docs/reference/labels-annotations-taints/_index.md (hunk 2 above)
Comment (sftim) [reply content/en/docs/reference/labels-annotations-taints/_index.md:541]: I would put this advice in the ServiceAccount page, not here. And strongly encourage readers to go look there.

File: content/en/docs/concepts/security/service-accounts.md (hunk 5 above)
Comment (sftim) [reply content/en/docs/concepts/security/service-accounts.md:220]: This looks wrong. How confident are you @chansuke that this is the actual behavior of Kubernetes?

File: content/en/docs/concepts/security/service-accounts.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/concepts/security/service-accounts.md:222]: I prefer the text @chansuke proposed, here. This link is to a reference page @tengqm and not to documentation about using the annotation.

**Comment by chansuke** [comment]
@tengqm Yes, I'm working on it now.

File: content/en/docs/concepts/configuration/secret.md [hunk 9]
```diff
@@ -807,6 +807,13 @@ There may be Secrets for several Pods on the same node. However, only the
 Secrets that a Pod requests are potentially visible within its containers.
 Therefore, one Pod does not have access to the Secrets of another Pod.
 
+### Configure least-privilege access to Secrets
+
+To enhance the security measures around Secrets, Kubernetes provides an mechanism: you c an
```
Comment (tengqm) [inline content/en/docs/concepts/configuration/secret.md:812]:
```suggestion
To enhance the security measures around Secrets, Kubernetes provides an mechanism: you can
```

File: content/en/docs/concepts/configuration/secret.md [hunk 10]
```diff
@@ -807,6 +807,13 @@ There may be Secrets for several Pods on the same node. However, only the
...
+annotate a ServiceAccount as `kubernetes.io/enforce-mountable-secrets: "true"`.
+
+For more information, you can refer to the [documentation about this annotation](https://k8s.io/docs/concepts/security/service-accounts/#enforce-mountable-secrets).
```
Comment (tengqm) [inline content/en/docs/concepts/configuration/secret.md:815]:
```suggestion
For more information, you can refer to the [documentation about this annotation](/docs/concepts/security/service-accounts/#enforce-mountable-secrets).
```

File: content/en/docs/concepts/security/service-accounts.md [hunk 11]
```diff
@@ -196,6 +196,31 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+    kubernetes.io/enforce-mountable-secrets: "true"
+  name: my-serviceaccount
+  namespace: my-namespace
+```
+
+When this annotation is set to "true", the Kubernetes control plane ensures that
+the Secrets from this ServiceAccount are only mountable on the resource kinds
+specified in the annotation value.
+For more information about the authentication process, refer to
+[kubernetes.io/enforce-mountable-secrets {#enforce-mountable-secrets}](https://k8s.io/docs/concepts/security/service-accounts/#enforce-mountable-secrets).
```
Comment (tengqm) [inline content/en/docs/concepts/security/service-accounts.md:222]:
```suggestion
[kubernetes.io/enforce-mountable-secrets {#enforce-mountable-secrets}](/docs/concepts/security/service-accounts/#enforce-mountable-secrets).
```

File: content/en/docs/concepts/configuration/secret.md [hunk 12]
```diff
@@ -807,6 +807,13 @@ There may be Secrets for several Pods on the same node. However, only the
 Secrets that a Pod requests are potentially visible within its containers.
 Therefore, one Pod does not have access to the Secrets of another Pod.
 
+### Configure least-privilege access to Secrets
+
+To enhance the security measures around Secrets, Kubernetes provides an mechanism: you can
```
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md:812]: (nit)
```suggestion
To enhance the security measures around Secrets, Kubernetes provides a mechanism: you can
```

File: content/en/docs/concepts/security/service-accounts.md [hunk 13]
```diff
@@ -196,6 +196,31 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+    kubernetes.io/enforce-mountable-secrets: "true"
+  name: my-serviceaccount
+  namespace: my-namespace
+```
+
+When this annotation is set to "true", the Kubernetes control plane ensures that
+the Secrets from this ServiceAccount are only mountable on the resource kinds
+specified in the annotation value.
+For more information about the authentication process, refer to
+[kubernetes.io/enforce-mountable-secrets {#enforce-mountable-secrets}](/docs/concepts/security/service-accounts/#enforce-mountable-secrets).
```
Comment (sftim) [inline content/en/docs/concepts/security/service-accounts.md:222]:
```suggestion
```

We shouldn't try to link to the section we're in.

File: content/en/docs/concepts/security/service-accounts.md [hunk 14]
```diff
@@ -196,6 +196,31 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+metadata:
+  annotations:
+    kubernetes.io/enforce-mountable-secrets: "true"
+  name: my-serviceaccount
+  namespace: my-namespace
+```
+
+When this annotation is set to "true", the Kubernetes control plane ensures that
+the Secrets from this ServiceAccount are only mountable on the resource kinds
+specified in the annotation value.
```
Comment (sftim) [inline content/en/docs/concepts/security/service-accounts.md:220]: This is technically inaccurate. If the annotation value is `true` it can't also be a list of resource kinds.

**Review by sftim — CHANGES_REQUESTED** [review]
/lgtm cancel

This PR seems to include hallucinations or guesses. Please write documentation that is internally consistent.

File: content/en/docs/concepts/security/secrets-good-practices.md [hunk 15]
```diff
@@ -62,6 +62,12 @@ recommendations include:
 *  Implement audit rules that alert on specific events, such as concurrent
    reading of multiple Secrets by a single user
 
+#### Additional ServiceAccount Annotations for Secret Management
```
Comment (sftim) [inline content/en/docs/concepts/security/secrets-good-practices.md:65]:
```suggestion
#### Additional ServiceAccount annotations for Secret management
```

File: content/en/docs/concepts/security/service-accounts.md [hunk 16]
```diff
@@ -196,6 +196,27 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+kind: ServiceAccount
+metadata:
+  annotations:
+    kubernetes.io/enforce-mountable-secrets: "true"
+  name: my-serviceaccount
+  namespace: my-namespace
+```
+When this annotation is set to "true", the Kubernetes control plane ensures that
+the Secrets from this ServiceAccount are subject to certain mounting restrictions.
+
```
Comment (sftim) [inline content/en/docs/concepts/security/service-accounts.md:219]: Ideally: document those restrictions.

File: content/en/docs/concepts/security/service-accounts.md [hunk 17]
```diff
@@ -196,6 +196,33 @@ or using a custom mechanism such as an [authentication webhook](/docs/reference/
...
+1. Secrets mounted as volumes must appear in the ServiceAccount's secrets field.
+1. Secrets referenced in envFrom in containers or initContainers must also appear in the ServiceAccount's secrets field.
+1. Secrets referenced in imagePullSecrets of Pods must match those in the ServiceAccount's imagePullSecrets field.
```
Comment (sftim) [inline content/en/docs/concepts/security/service-accounts.md:222]: Something like this?
```suggestion
1. The name of each Secret that is mounted as a volume in a Pod must appear in the `secrets` field of the
   Pod's ServiceAccount.
1. The name of each Secret referenced using `envFrom` in a Pod must also appear in the `secrets`
   field of the Pod's ServiceAccount.
1. The name of each Secret referenced using `imagePullSecrets` in a Pod must also appear in the `secrets`
   field of the Pod's ServiceAccount.
```
or, better, try to simplify the wording.

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 18]
```diff
@@ -845,6 +845,12 @@ The Kubernetes project strongly recommends enabling this admission controller.
 You should enable this admission controller if you intend to make any use of Kubernetes
 `ServiceAccount` objects.
 
+Regarding the annotation `kubernetes.io/enforce-mountable-secrets`: While the annotation's name suggests it only concerns the mounting of Secrets,
+its enforcement also extends to other ways Secrets are used in the context of a Pod.
+Therefore, it is crucial to ensure that all the referenced secrets are correctly specified in the ServiceAccount.
+
+For more details, see the [documentation on this annotation](/docs/reference/labels-annotations-taints/#enforce-mountable-secrets).
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:852]: This is too much detail for this page, I'm afraid.

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 19]
```diff
@@ -530,8 +530,13 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
 Used on: ServiceAccount
 
 The value for this annotation must be **true** to take effect.
-This annotation indicates that Pods running as this ServiceAccount may only reference
-Secret API objects specified in the ServiceAccount's `secrets` field.
+When set to "true", this annotation enforces the following rules for Pods running as this ServiceAccount:
```
Comment (sftim) [inline content/en/docs/reference/labels-annotations-taints/_index.md:533]:
```suggestion
When you set this annotation  to "true", Kubernetes enforces the following rules for
Pods running as this ServiceAccount:
```

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 20]
```diff
@@ -530,8 +530,13 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
...
+
+1. Secrets mounted as volumes must be listed in the ServiceAccount's `secrets` field.
+1. Secrets referenced in `envFrom` for containers or `initContainers` must also
+   be listed in the ServiceAccount's `secrets` field.
+1. Secrets referenced in a Pod's `imagePullSecrets` must be present in the
+   ServiceAccount's `imagePullSecrets` field.
```
Comment (sftim) [inline content/en/docs/reference/labels-annotations-taints/_index.md:540]: Maybe state what happens to a Pod that doesn't meet those requirements, and also when the enforcement happens.

**Review by sftim — COMMENTED** [review]
A bit more feedback. This is complicated but also important to get right.

**Comment by chansuke** [comment]
@sftim I've updated, please review:pray:

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 21]
```diff
@@ -548,8 +548,18 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
...
 
 The value for this annotation must be **true** to take effect.
-This annotation indicates that Pods running as this ServiceAccount may only reference
-Secret API objects specified in the ServiceAccount's `secrets` field.
+When you set this annotation  to "true", Kubernetes enforces the following rules for
+Pods running as this ServiceAccount:
+
+1. Secrets mounted as volumes must be listed in the ServiceAccount's `secrets` field.
+1. Secrets referenced in envFrom for containers or initContainers must also be listed in the ServiceAccount's secrets field.
+   If an initContainer references a secret not listed in the ServiceAccount's secrets field,
```
Comment (sftim) [inline content/en/docs/reference/labels-annotations-taints/_index.md:556]: Is this right?
```suggestion
   If any container in a Pod references a Secret not listed in the ServiceAccount's `secrets` field
   (and even if the reference is marked as `optional`), then
```

File: content/en/docs/reference/labels-annotations-taints/_index.md [hunk 22]
```diff
@@ -548,8 +548,18 @@ Example: `kubernetes.io/enforce-mountable-secrets: "true"`
...
+   the Pod will fail to start, and an error indicating the non-compliant secret reference will be generated.
+1. Secrets referenced in a Pod's `imagePullSecrets` must be present in the
+   ServiceAccount's `imagePullSecrets` field, the Pod will fail to start,
+   and an error indicating the non-compliant image pull secret reference will be generated.
+
+When you create or update a Pod, these rules are checked. If a Pod doesn't follow them, it won't start and you'll see an error message.
```
Comment (sftim) [inline content/en/docs/reference/labels-annotations-taints/_index.md:563]:
```suggestion
If a Pod is already running and you change the `kubernetes.io/enforce-mountable-secrets` annotation
to true, or you edit the associated ServiceAccount to remove the reference to a Secret
that the Pod is already using, the Pod continues to run.

```
(is this right?)

**Comment by niranjandarshann** [comment]
@chansuke  
Could you please squash your commits: https://kubernetes.io/docs/contribute/new-content/open-a-pr/#squashing-commits

**Comment by sftim** [comment]
I'm happy with the changes since https://github.com/kubernetes/website/pull/42741#issuecomment-1694634155

/lgtm
/remove-label tide/merge-method-squash

**Review by sftim — COMMENTED** [review]
/approve

Thank you for this particularly helpful PR, @chansuke

**Comment by chansuke** [comment]
@sftim Thank you so much for your support.
